MariaDB Replication Is Broken

### Description of the problem

when we use MariaDB replication with Gramine it fails with the following error

```bash
Relay log read failure: Could not parse relay log event entry. The possible reasons are: the master's binary log is corrupted (you can check this by running 'mysqlbinlog' on the binary log), the slave's relay log is corrupted (you can check this by running 'mysqlbinlog' on the relay log), a network problem, or a bug in the master's or slave's MariaDB code. If you want to check the master's binary log or slave's relay log, you will be able to know their names by issuing 'SHOW SLAVE STATUS' on this slave
``` 

Without Gramine all is good, but with Gramine we have this error. This is not also related to the sgx because we get the same error with gramine-direct as well.

### Steps to reproduce

I used this https://github.com/enclaive/enclaive-docker-mariadb-sgx repository to reproduce this issue but the issue also exists on the Gramine 1.5,1.6 and 1.7 versions as well.

Clone the repository

```bash
git pull https://github.com/enclaive/enclaive-docker-mariadb-sgx.git
```

Build docker image

```bash
docker build -t gramine-mariadb:v2 .
```

then create 2 instances with this docker image and create replication between them.

```bash 
docker run -it --name master --security-opt seccomp=/root/docker_seccomp_aug_2022.json gramine-mariadb:v2
docker run -it --name replica --security-opt seccomp=/root/docker_seccomp_aug_2022.json gramine-mariadb:v2

CHANGE MASTER TO MASTER_HOST='<master-ip>', MASTER_USER='<user>',MASTER_PASSWORD='<password>',MASTER_USE_GTID = slave_pos;
```

Once the replication is started all is good. But some time later it fails with this error. You can just run the following sysbench test and reproduce this replication error

```bash
sysbench oltp_read_only --mysql-host=127.0.0.1 --tables=10 --table_size=50000  --mysql-port=3306 --mysql-user=<user>--mysql-password="<password>" --mysql-db=sbtest  prepare
```



### Expected results

A running replication stream

### Actual results

replication status

```bash
MariaDB [sbtest]> show slave status \G
*************************** 1. row ***************************
                Slave_IO_State:
                   Master_Host: 172.17.0.2
                   Master_User: root
                   Master_Port: 3306
                 Connect_Retry: 60
               Master_Log_File: ON.000001
           Read_Master_Log_Pos: 7060656
                Relay_Log_File: mysqld-relay-bin.000002
                 Relay_Log_Pos: 5487680
         Relay_Master_Log_File: ON.000001
              Slave_IO_Running: No
             Slave_SQL_Running: No
               Replicate_Do_DB:
           Replicate_Ignore_DB:
            Replicate_Do_Table:
        Replicate_Ignore_Table:
       Replicate_Wild_Do_Table:
   Replicate_Wild_Ignore_Table:
                    Last_Errno: 1594
                    Last_Error: Relay log read failure: Could not parse relay log event entry. The possible reasons are: the master's binary log is corrupted (you can check this by running 'mysqlbinlog' on the binary log), the slave's relay log is corrupted (you can check this by running 'mysqlbinlog' on the relay log), a network problem, or a bug in the master's or slave's MariaDB code. If you want to check the master's binary log or slave's relay log, you will be able to know their names by issuing 'SHOW SLAVE STATUS' on this slave.
                  Skip_Counter: 0
           Exec_Master_Log_Pos: 5487388
               Relay_Log_Space: 7061258
               Until_Condition: None
                Until_Log_File:
                 Until_Log_Pos: 0
            Master_SSL_Allowed: No
            Master_SSL_CA_File:
            Master_SSL_CA_Path:
               Master_SSL_Cert:
             Master_SSL_Cipher:
                Master_SSL_Key:
         Seconds_Behind_Master: NULL
 Master_SSL_Verify_Server_Cert: No
                 Last_IO_Errno: 1595
                 Last_IO_Error: Relay log write failure: could not queue event from master
                Last_SQL_Errno: 1594
                Last_SQL_Error: Relay log read failure: Could not parse relay log event entry. The possible reasons are: the master's binary log is corrupted (you can check this by running 'mysqlbinlog' on the binary log), the slave's relay log is corrupted (you can check this by running 'mysqlbinlog' on the relay log), a network problem, or a bug in the master's or slave's MariaDB code. If you want to check the master's binary log or slave's relay log, you will be able to know their names by issuing 'SHOW SLAVE STATUS' on this slave.
   Replicate_Ignore_Server_Ids:
              Master_Server_Id: 1
                Master_SSL_Crl:
            Master_SSL_Crlpath:
                    Using_Gtid: Slave_Pos
                   Gtid_IO_Pos: 0-1-33
       Replicate_Do_Domain_Ids:
   Replicate_Ignore_Domain_Ids:
                 Parallel_Mode: optimistic
                     SQL_Delay: 0
           SQL_Remaining_Delay: NULL
       Slave_SQL_Running_State:
              Slave_DDL_Groups: 14
Slave_Non_Transactional_Groups: 0
    Slave_Transactional_Groups: 17
````


if we check the relay logs with the following command we see they are corrupted.

```bash

 show relaylog events IN 'mysqld-relay-bin.000002';
ERROR 1220 (HY000): Error when executing command SHOW BINLOG EVENTS: Wrong offset or I/O error



 show relaylog events IN 'mysqld-relay-bin.000002' limit 10,3;
+-------------------------+------+------------+-----------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Log_name                | Pos  | Event_type | Server_id | End_log_pos | Info                                                                                                                                                                                                                                                                                                                                                      |
+-------------------------+------+------------+-----------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| mysqld-relay-bin.000002 |  961 | Query      |         1 |         756 | create database test                                                                                                                                                                                                                                                                                                                                      |
| mysqld-relay-bin.000002 | 1048 | Gtid       |         1 |         798 | GTID 0-1-4                                                                                                                                                                                                                                                                                                                                                |
| mysqld-relay-bin.000002 | 1090 | Query      |         1 |        1207 | use `test`; CREATE TABLE test_table (
    id INT AUTO_INCREMENT PRIMARY KEY,
    name VARCHAR(255) NOT NULL,
    email VARCHAR(255) NOT NULL,
    phone VARCHAR(20),
    address VARCHAR(255),
    city VARCHAR(100),
    state VARCHAR(100),
    postal_code VARCHAR(20),
    country VARCHAR(100),
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
) |
+-------------------------+------+------------+-----------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
3 rows in set (0.000 sec)
```

logs:
```bash
2024-06-24 11:51:51 9 [Note] Slave I/O thread: connected to master 'root@172.17.0.2:3306',replication starts at GTID position ''
2024-06-24 13:32:14 10 [ERROR] Error in Log_event::read_log_event(): 'Replication event checksum verification failed while reading from a log file', data_len: 524293, event_type: 2
2024-06-24 13:32:14 9 [ERROR] Slave I/O: Relay log write failure: could not queue event from master, Internal MariaDB error code: 1595
2024-06-24 13:32:14 9 [Note] Slave I/O thread exiting, read up to log 'ON.000001', position 7060656; GTID position 0-1-33, master 172.17.0.2:3306
2024-06-24 13:32:14 10 [ERROR] Error reading relay log event: slave SQL thread aborted because of I/O error
2024-06-24 13:32:14 10 [ERROR] Slave SQL: Relay log read failure: Could not parse relay log event entry. The possible reasons are: the master's binary log is corrupted (you can check this by running 'mysqlbinlog' on the binary log), the slave's relay log is corrupted (you can check this by running 'mysqlbinlog' on the relay log), a network problem, or a bug in the master's or slave's MariaDB code. If you want to check the master's binary log or slave's relay log, you will be able to know their names by issuing 'SHOW SLAVE STATUS' on this slave. Internal MariaDB error code: 1594
2024-06-24 13:32:14 10 [ERROR] Error running query, slave SQL thread aborted. Fix the problem, and restart the slave SQL thread with "SLAVE START". We stopped at log 'ON.000001' position 5487388; GTID position '0-1-30'
2024-06-24 13:32:14 10 [Note] Slave SQL thread exiting, replication stopped in log 'ON.000001' at position 5487388; GTID position '0-1-30', master: 172.17.0.2:3306
2024-06-24 21:32:43 11 [Warning] Aborted connection 11 to db: 'sbtest' user: 'root' host: '127.0.0.1' (Got timeout reading communication packets)
```


### Gramine commit hash

10e93534169802be16fc9e2b3e9ac70d08efcb41

@dimakuv, the MariaDB replication has 2 threads. the first one is responsible for downloading the binary log files from the master (IO_THREAD) and the other one(SQL_THREAD) is responsible for reading the binary logs from this relay log file and applying them to the replica database.

if I set up replication and stop SQL_THREAD, the replica downloads the binary logs without issues. If I stop the IO_THREAD and start SQL_THREAD, the SQL_THREAD applies the downloaded binary logs without issues. But if I let them run at the same time it ends up with the replication error that I shared before.

I compiled mariadb with debug and reproduced the issue. That time it crashed with the following trace.

```bash
mariadbd: /server/mysys/mf_iocache.c:1723: my_b_flush_io_cache: Assertion `info->end_of_file == inline_mysql_file_tell("/server/mysys/mf_iocache.c", 1723, info->file, (myf) (0))' failed.
240703 11:51:27 [ERROR] mysqld got signal 6 ;
Sorry, we probably made a mistake, and this is a bug.

Your assistance in bug reporting will enable us to fix this for the next release.
To report this bug, see https://mariadb.com/kb/en/reporting-bugs

We will try our best to scrape up some info that will hopefully help
diagnose the problem, but since we have already crashed,
something is definitely wrong and this may fail.

Server version: 10.6.19-MariaDB-debug-log source revision: f6fcfc1a6a058fd7cac6bf53216ea73f3a04b22d
key_buffer_size=134217728
read_buffer_size=131072
max_used_connections=1
max_threads=153
thread_count=3
It is possible that mysqld could use up to
key_buffer_size + (read_buffer_size + sort_buffer_size)*max_threads = 468166 K  bytes of memory
Hope that's ok; if not, decrease some variables in the equation.

Thread pointer: 0x24582abdb098
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
stack_bottom = 0x24582abd9b48 thread_stack 0x49000
/app/mariadbd(my_print_stacktrace+0x44)[0x2458481fa5e9]
/app/mariadbd(handle_fatal_signal+0x421)[0x2458477d7b76]
libc_sigaction.c:0(__restore_rt)[0x2458461b8ed0]
pthread_kill.c:0(__pthread_kill_implementation)[0x245846204766]
/lib/libc.so.6(raise+0x16)[0x2458461b8e26]
/lib/libc.so.6(abort+0xc5)[0x2458461a37e5]
loadmsgcat.c:0(_nl_load_domain.cold)[0x2458461a371b]
/lib/libc.so.6(+0x3acc6)[0x2458461b1cc6]
/app/mariadbd(my_b_flush_io_cache+0x2a4)[0x2458481ce3ec]
/app/mariadbd(_ZN13MYSQL_BIN_LOG14flush_and_syncEPb+0x110)[0x245847937cc2]
/app/mariadbd(_ZN13MYSQL_BIN_LOG18write_event_bufferEPhj+0x509)[0x245847937ad5]
/app/mariadbd(+0x8f4353)[0x2458472f1353]
/app/mariadbd(handle_slave_io+0x13c2)[0x2458472e9f0f]
/app/mariadbd(+0x1198ccb)[0x245847b95ccb]
pthread_create.c:0(start_thread)[0x245846202a17]
/lib/libc.so.6(clone+0x51)[0x245846284d61]

Trying to get some variables.
Some pointers may be invalid and cause the dump to abort.
Query (0x0): (null)
Connection ID (thread ID): 6
Status: NOT_KILLED

Optimizer switch: index_merge=on,index_merge_union=on,index_merge_sort_union=on,index_merge_intersection=on,index_merge_sort_intersection=off,engine_condition_pushdown=off,index_condition_pushdown=on,derived_merge=on,derived_with_keys=on,firstmatch=on,loosescan=on,materialization=on,in_to_exists=on,semijoin=on,partial_match_rowid_merge=on,partial_match_table_scan=on,subquery_cache=on,mrr=off,mrr_cost_based=off,mrr_sort_keys=off,outer_join_with_cache=on,semijoin_with_cache=on,join_cache_incremental=on,join_cache_hashed=on,join_cache_bka=on,optimize_join_buffer_size=on,table_elimination=on,extended_keys=on,exists_to_in=on,orderby_uses_equalities=on,condition_pushdown_for_derived=on,split_materialized=on,condition_pushdown_for_subquery=on,rowid_filter=on,condition_pushdown_from_having=on,not_null_range_scan=off,hash_join_cardinality=off,cset_narrowing=off

The manual page at https://mariadb.com/kb/en/how-to-produce-a-full-stack-trace-for-mariadbd/ contains
information that should help you find out what is causing the crash.

```

so it points this part of the MariaDB server code;

https://github.com/MariaDB/server/blob/7c5fdc9b6a45f4e8bad1ae5c7c93b457be9891fa/mysys/mf_iocache.c#L1804-L1808

```bash
        info->end_of_file= eof;
        info->append_read_pos= info->write_buffer;
        DBUG_ASSERT(info->end_of_file == mysql_file_tell(info->file, MYF(0)));
      }
      else
```

it is in `mf_iocache.c` and `my_b_flush_io_cache` function.

it has the following comments section of this file [here](https://github.com/MariaDB/server/blob/11.6/mysys/mf_iocache.c#L1804-L1808),

> /*
  Cashing of files with only does (sequential) read or writes of fixed-
  length records. A read isn't allowed to go over file-length. A read is ok
  if it ends at file-length and next read can try to read after file-length
  (and get a EOF-error).
  Possibly use of asyncronic io.
  macros for read and writes for faster io.
  Used instead of FILE when reading or writing whole files.
  One can change info->pos_in_file to a higher value to skip bytes in file if
  also info->read_pos is set to info->read_end.
  If called through open_cached_file(), then the temporary file will
  only be created if a write exeeds the file buffer or if one calls
  my_b_flush_io_cache().

>  If one uses SEQ_READ_APPEND, then two buffers are allocated, one for
  reading and another for writing.  Reads are first done from disk and
  then done from the write buffer.  This is an efficient way to read
  from a log file when one is writing to it at the same time.
  For this to work, the file has to be opened in append mode!
  Note that when one uses SEQ_READ_APPEND, one MUST write using
  my_b_append !  This is needed because we need to lock the mutex
  every time we access the write buffer.

>TODO:
  When one SEQ_READ_APPEND and we are reading and writing at the same time,
  each time the write buffer gets full and it's written to disk, we will
  always do a disk read to read a part of the buffer from disk to the
  read buffer.
  This should be fixed so that when we do a my_b_flush_io_cache() and
  we have been reading the write buffer, we should transfer the rest of the
  write buffer to the read buffer before we start to reuse it.
*/

Basically, there is a writer thread and a reader thread. They both use the same file. There is also a buffer, the reader thread first reads from the disk and then reads the buffer. Maybe reading the buffer with Gramine causes some conflicts.

I will keep reading the code, but please note that I am not a c developer and an expert on that topic. But will try to find some more details.

I also set the Gramine log level to trace and re-produced the issue again. Sending the logs in the attachment.
[replica_logs.tar.gz](https://github.com/user-attachments/files/16085447/replica_logs.tar.gz)

Thanks @dimakuv , I signed the image and I made it work. Unfortunately, the same issue still exists. I will enable trace logging and send you gramine trace again.

```bash
2024-07-05 14:11:27 8 [ERROR] Error in Log_event::read_log_event(): 'Replication event checksum verification failed while reading from a log file', data_len: 524293, event_type: 2
2024-07-05 14:11:27 8 [ERROR] Error reading relay log event: slave SQL thread aborted because of I/O error
2024-07-05 14:11:27 8 [ERROR] Slave SQL: Relay log read failure: Could not parse relay log event entry. The possible reasons are: the master's binary log is corrupted (you can check this by running 'mysqlbinlog' on the binary log), the slave's relay log is corrupted (you can check this by running 'mysqlbinlog' on the relay log), a network problem, or a bug in the master's or slave's MariaDB code. If you want to check the master's binary log or slave's relay log, you will be able to know their names by issuing 'SHOW SLAVE STATUS' on this slave. Internal MariaDB error code: 1594
2024-07-05 14:11:27 8 [ERROR] Error running query, slave SQL thread aborted. Fix the problem, and restart the slave SQL thread with "SLAVE START". We stopped at log 'ON.000001' position 26227194; GTID position '0-1-59'
2024-07-05 14:11:27 8 [Note] Slave SQL thread exiting, replication stopped in log 'ON.000001' at position 26227194; GTID position '0-1-59', master: 172.30.43.5:3306
2024-07-05 14:11:27 7 [ERROR] Slave I/O: Relay log write failure: could not queue event from master, Internal MariaDB error code: 1595
2024-07-05 14:11:27 7 [Note] Slave I/O thread exiting, read up to log 'ON.000001', position 26751666; GTID position 0-1-60, master 172.30.43.5:3306
```

```bash 
*************************** 1. row ***************************
                Slave_IO_State:
                   Master_Host: 172.30.43.5
                   Master_User: root
                   Master_Port: 3306
                 Connect_Retry: 60
               Master_Log_File: ON.000001
           Read_Master_Log_Pos: 26751666
                Relay_Log_File: mysqld-relay-bin.000002
                 Relay_Log_Pos: 26227486
         Relay_Master_Log_File: ON.000001
              Slave_IO_Running: No
             Slave_SQL_Running: No
               Replicate_Do_DB:
           Replicate_Ignore_DB:
            Replicate_Do_Table:
        Replicate_Ignore_Table:
       Replicate_Wild_Do_Table:
   Replicate_Wild_Ignore_Table:
                    Last_Errno: 1594
                    Last_Error: Relay log read failure: Could not parse relay log event entry. The possible reasons are: the master's binary log is corrupted (you can check this by running 'mysqlbinlog' on the binary log), the slave's relay log is corrupted (you can check this by running 'mysqlbinlog' on the relay log), a network problem, or a bug in the master's or slave's MariaDB code. If you want to check the master's binary log or slave's relay log, you will be able to know their names by issuing 'SHOW SLAVE STATUS' on this slave.
                  Skip_Counter: 0
           Exec_Master_Log_Pos: 26227194
               Relay_Log_Space: 26752268
               Until_Condition: None
                Until_Log_File:
                 Until_Log_Pos: 0
            Master_SSL_Allowed: No
            Master_SSL_CA_File:
            Master_SSL_CA_Path:
               Master_SSL_Cert:
             Master_SSL_Cipher:
                Master_SSL_Key:
         Seconds_Behind_Master: NULL
 Master_SSL_Verify_Server_Cert: No
                 Last_IO_Errno: 1595
                 Last_IO_Error: Relay log write failure: could not queue event from master
                Last_SQL_Errno: 1594
                Last_SQL_Error: Relay log read failure: Could not parse relay log event entry. The possible reasons are: the master's binary log is corrupted (you can check this by running 'mysqlbinlog' on the binary log), the slave's relay log is corrupted (you can check this by running 'mysqlbinlog' on the relay log), a network problem, or a bug in the master's or slave's MariaDB code. If you want to check the master's binary log or slave's relay log, you will be able to know their names by issuing 'SHOW SLAVE STATUS' on this slave.
   Replicate_Ignore_Server_Ids:
              Master_Server_Id: 1
                Master_SSL_Crl:
            Master_SSL_Crlpath:
                    Using_Gtid: Slave_Pos
                   Gtid_IO_Pos: 0-1-60
       Replicate_Do_Domain_Ids:
   Replicate_Ignore_Domain_Ids:
                 Parallel_Mode: optimistic
                     SQL_Delay: 0
           SQL_Remaining_Delay: NULL
       Slave_SQL_Running_State:
              Slave_DDL_Groups: 8
Slave_Non_Transactional_Groups: 0
    Slave_Transactional_Groups: 52
```

--- a/Documentation/devel/features.md
+++ b/Documentation/devel/features.md
@@ -2058,8 +2058,10 @@ Gramine implements all classic filesystem operations, but with limitations descr
 
 Gramine supports opening files and directories (via `open()` and `openat()` system calls).
 `O_CLOEXEC`, `O_CREAT`, `O_DIRECTORY`, `O_EXCL`, `O_NOFOLLOW`, `O_PATH`, `O_TRUNC` flags are
-supported. Other flags are ignored. Notable ignored flags are `O_APPEND` (not yet implemented in
-Gramine) and `O_TMPFILE` (bug in Gramine: should not be silently ignored).
+supported. `O_APPEND` is supported, but is currently limited only to a single process, i.e., opening
+the same file in append mode in two processes may result in file contents being overwritten. Other
+flags are ignored. Notable ignored flag is `O_TMPFILE` (bug in Gramine: should not be silently
+ignored).
 
 Trusted files can be opened only for reading. Already-existing encrypted files can be opened only if
 they were not moved or renamed on the host (this is for protection against file renaming attacks).

--- a/libos/include/libos_fs.h
+++ b/libos/include/libos_fs.h
@@ -726,6 +726,7 @@ int get_dirfd_dentry(int dirfd, struct libos_dentry** dir);
  * - O_DIRECTORY: expect/create a directory instead of regular file
  * - O_NOFOLLOW: don't follow symbolic links when resolving a path
  * - O_TRUNC: truncate the file after opening
+ * - O_APPEND: open file in append mode for writing
  *
  * The flags (including any not listed above), as well as file mode, are passed to the underlying
  * filesystem.

--- a/libos/src/fs/chroot/encrypted.c
+++ b/libos/src/fs/chroot/encrypted.c
@@ -477,14 +477,15 @@ static ssize_t chroot_encrypted_write(struct libos_handle* hdl, const void* buf,
 
     lock(&hdl->inode->lock);
 
-    int ret = encrypted_file_write(enc, buf, count, *pos, &actual_count);
+    file_off_t actual_pos = (hdl->flags & O_APPEND) ? hdl->inode->size : *pos;
+    int ret = encrypted_file_write(enc, buf, count, actual_pos, &actual_count);
     if (ret < 0) {
         unlock(&hdl->inode->lock);
         return ret;
     }
 
     assert(actual_count <= count);
-    *pos += actual_count;
+    *pos = actual_pos + actual_count;
     if (hdl->inode->size < *pos)
         hdl->inode->size = *pos;
 

--- a/libos/src/fs/chroot/fs.c
+++ b/libos/src/fs/chroot/fs.c
@@ -623,14 +623,20 @@ static ssize_t chroot_write(struct libos_handle* hdl, const void* buf, size_t co
         return -EACCES;
     }
 
-    int ret = PalStreamWrite(hdl->pal_handle, *pos, &count, (void*)buf);
+    file_off_t actual_pos = *pos;
+    lock(&hdl->inode->lock);
+    if (hdl->inode->type == S_IFREG && (hdl->flags & O_APPEND))
+        actual_pos = hdl->inode->size;
+    unlock(&hdl->inode->lock);
+
+    int ret = PalStreamWrite(hdl->pal_handle, actual_pos, &count, (void*)buf);
     if (ret < 0) {
         return pal_to_unix_errno(ret);
     }
 
     size_t new_size = 0;
     if (hdl->inode->type == S_IFREG) {
-        *pos += count;
+        *pos = actual_pos + count;
         /* Update file size if we just wrote past the end of file */
         lock(&hdl->inode->lock);
         if (hdl->inode->size < *pos)

--- a/libos/src/fs/tmpfs/fs.c
+++ b/libos/src/fs/tmpfs/fs.c
@@ -259,7 +259,8 @@ static ssize_t tmpfs_write(struct libos_handle* hdl, const void* buf, size_t siz
     lock(&inode->lock);
     struct libos_mem_file* mem = inode->data;
 
-    ret = mem_file_write(mem, *pos, buf, size);
+    file_off_t actual_pos = (hdl->flags & O_APPEND) ? hdl->inode->size : *pos;
+    ret = mem_file_write(mem, actual_pos, buf, size);
     if (ret < 0) {
         unlock(&inode->lock);
         return ret;
@@ -267,7 +268,7 @@ static ssize_t tmpfs_write(struct libos_handle* hdl, const void* buf, size_t siz
 
     inode->size = mem->size;
 
-    *pos += ret;
+    *pos = actual_pos + ret;
     inode->mtime = time_us / USEC_IN_SEC;
     /* keep `ret` */
 

--- a/libos/src/sys/libos_file.c
+++ b/libos/src/sys/libos_file.c
@@ -442,6 +442,12 @@ long libos_syscall_sendfile(int out_fd, int in_fd, off_t* offset, size_t count) 
         goto out;
     }
 
+    if (!(in_hdl->acc_mode & MAY_READ) || !(out_hdl->acc_mode & MAY_WRITE)) {
+        /* error out if input fd isn't readable or output fd isn't writable */
+        ret = -EBADF;
+        goto out;
+    }
+
     if (out_hdl->flags & O_APPEND) {
         /* Linux errors out if output fd has the O_APPEND flag set; comply with this behavior */
         ret = -EINVAL;
@@ -491,12 +497,6 @@ long libos_syscall_sendfile(int out_fd, int in_fd, off_t* offset, size_t count) 
         maybe_lock_pos_handle(in_hdl);
         pos_in = in_hdl->pos;
         maybe_unlock_pos_handle(in_hdl);
-    }
-
-    if (!(out_hdl->acc_mode & MAY_WRITE)) {
-        /* Linux errors out if output fd isn't writable */
-        ret = -EBADF;
-        goto out;
     }
 
     while (copied_to_out < count) {

--- a/libos/src/sys/libos_open.c
+++ b/libos/src/sys/libos_open.c
@@ -330,6 +330,14 @@ long libos_syscall_pwrite64(int fd, char* buf, size_t count, loff_t offset) {
         goto out;
     }
 
+    /*
+     * Note that Linux has the following bug: POSIX requires that opening a file with the O_APPEND
+     * flag should have no effect on the location at which pwrite() writes data. However, on Linux,
+     * if a file is opened with O_APPEND, pwrite() appends data to the end of the file, regardless
+     * of the value of offset. See man page pwrite(2), section BUGS.
+     *
+     * Gramine complies with this behavior, see implementations of the write() callback.
+     */
     file_off_t pos = offset;
     ret = fs->fs_ops->write(hdl, buf, count, &pos);
 out:

--- a/libos/test/fs/common.c
+++ b/libos/test/fs/common.c
@@ -65,6 +65,13 @@ int open_output_fd(const char* path, bool rdwr) {
     return fd;
 }
 
+int open_output_fd_append(const char* path) {
+    int fd = open(path, O_RDWR | O_CREAT | O_APPEND, 0664);
+    if (fd < 0)
+        fatal_error("Failed to open output file for append %s: %s\n", path, strerror(errno));
+    return fd;
+}
+
 void write_fd(const char* path, int fd, const void* buffer, size_t size) {
     off_t offset = 0;
     while (size > 0) {

--- a/libos/test/fs/common.h
+++ b/libos/test/fs/common.h
@@ -38,6 +38,7 @@ void read_fd(const char* path, int fd, void* buffer, size_t size);
 void seek_fd(const char* path, int fd, off_t offset, int mode);
 off_t tell_fd(const char* path, int fd);
 int open_output_fd(const char* path, bool rdwr);
+int open_output_fd_append(const char* path);
 void write_fd(const char* path, int fd, const void* buffer, size_t size);
 void sendfile_fd(const char* input_path, const char* output_path, int fi, int fo, size_t size);
 void close_fd(const char* path, int fd);

--- a/libos/test/fs/meson.build
+++ b/libos/test/fs/meson.build
@@ -42,6 +42,7 @@ tests = {
     },
     'open_close': {},
     'open_flags': {},
+    'read_append': {},
     'read_write': {},
     'read_write_mmap': {},
     'seek_tell': {},

--- a/libos/test/fs/read_append.c
+++ b/libos/test/fs/read_append.c
@@ -1,0 +1,124 @@
+#include "common.h"
+
+static void read_append(const char* file_path) {
+    const size_t size = 1024 * 1024;
+
+    void* buf1 = alloc_buffer(size);
+    void* buf2 = alloc_buffer(size);
+    fill_random(buf1, size);
+
+    ssize_t bytes_read;
+
+    /* test 1: create new file, append buf1 in two chunks to it, try to read (this should return EOF
+     * because appends moved file pos), then reset the file pos and read again */
+    printf("TEST 1 (%s)\n", file_path);
+    int fd = open_output_fd_append(file_path);
+    printf("open(%s) OK\n", file_path);
+    write_fd(file_path, fd, buf1, size / 2);
+    printf("first write(%s) OK\n", file_path);
+    write_fd(file_path, fd, buf1 + size / 2, size - size / 2);
+    printf("second write(%s) OK\n", file_path);
+    bytes_read = read(fd, buf2, size);
+    if (bytes_read != 0)
+        fatal_error("Read after append did not indicate EOF\n");
+    printf("first read(%s) OK\n", file_path);
+    seek_fd(file_path, fd, 0, SEEK_SET);
+    printf("seek(%s) OK\n", file_path);
+    read_fd(file_path, fd, buf2, size);
+    printf("second read(%s) OK\n", file_path);
+    if (memcmp(buf1, buf2, size) != 0)
+        fatal_error("Read data is different from what was written\n");
+    printf("compare(%s) OK\n", file_path);
+    close_fd(file_path, fd);
+    printf("close(%s) OK\n", file_path);
+    size_t file_size1 = file_size(file_path);
+    if (file_size1 != size)
+        fatal_error("File size is wrong (expected %lu got %lu)\n", size, file_size1);
+    printf("file_size(%s) OK\n", file_path);
+
+
+    /* test 2: open the file created previously (it must contain buf1), then read (this should
+     * return buf1 as there was no write/append operation yet, so file pos is zero), then reset the
+     * file pos and append buf1 (this should result in buf1 + buf1 contents of the file) */
+    printf("TEST 2 (%s)\n", file_path);
+    fd = open_output_fd_append(file_path);
+    printf("open(%s) OK\n", file_path);
+    read_fd(file_path, fd, buf2, size);
+    printf("first read(%s) OK\n", file_path);
+    if (memcmp(buf1, buf2, size) != 0)
+        fatal_error("Read data is different from what was written\n");
+    printf("first compare(%s) OK\n", file_path);
+    bytes_read = read(fd, buf2, size);
+    if (bytes_read != 0)
+        fatal_error("Second read did not indicate EOF\n");
+    printf("second read(%s) OK\n", file_path);
+    seek_fd(file_path, fd, 0, SEEK_SET);
+    printf("seek(%s) OK\n", file_path);
+    write_fd(file_path, fd, buf1, size);
+    printf("write(%s) OK\n", file_path);
+    seek_fd(file_path, fd, 0, SEEK_SET);
+    printf("seek(%s) OK\n", file_path);
+    read_fd(file_path, fd, buf2, size);
+    printf("first read(%s) after appending OK\n", file_path);
+    if (memcmp(buf1, buf2, size) != 0)
+        fatal_error("Read data is different from what was written\n");
+    read_fd(file_path, fd, buf2, size);
+    printf("second read(%s) after appending OK\n", file_path);
+    if (memcmp(buf1, buf2, size) != 0)
+        fatal_error("Read data is different from what was written\n");
+    printf("compare(%s) OK\n", file_path);
+    close_fd(file_path, fd);
+    printf("close(%s) OK\n", file_path);
+    size_t file_size2 = file_size(file_path);
+    if (file_size2 != size * 2)
+        fatal_error("File size is wrong (expected %lu got %lu)\n", size * 2, file_size2);
+    printf("file_size(%s) OK\n", file_path);
+
+    /* test 3: open the file created previously (size must be sizeof(buf1) * 2) in no-append mode,
+     * reset file pos and write buf1 (must overwrite contents, so file size must stay the same),
+     * then change to append mode, reset file pos and write buf1 (must append, so size will
+     * increase) */
+    printf("TEST 3 (%s)\n", file_path);
+    fd = open_output_fd(file_path, /*rdwr=*/true);
+    printf("open(%s) OK\n", file_path);
+    seek_fd(file_path, fd, 0, SEEK_SET);
+    printf("first seek(%s) OK\n", file_path);
+    write_fd(file_path, fd, buf1, size);
+    printf("first write(%s) OK\n", file_path);
+    int fcntl_ret = fcntl(fd, F_SETFL, O_APPEND);
+    if (fcntl_ret < 0)
+        fatal_error("Could not set O_APPEND flag using fcntl()\n");
+    seek_fd(file_path, fd, 0, SEEK_SET);
+    printf("second seek(%s) OK\n", file_path);
+    write_fd(file_path, fd, buf1, size);
+    printf("second write(%s) OK\n", file_path);
+    seek_fd(file_path, fd, 0, SEEK_SET);
+    printf("third seek(%s) OK\n", file_path);
+    for (int i = 0; i < 2; i++) {
+        read_fd(file_path, fd, buf2, size);
+        printf("read number %d (%s) OK\n", i + 1, file_path);
+        if (memcmp(buf1, buf2, size) != 0)
+            fatal_error("read number %d: Read data is different from what was written\n", i + 1);
+    }
+    printf("compare(%s) OK\n", file_path);
+    close_fd(file_path, fd);
+    printf("close(%s) OK\n", file_path);
+    size_t file_size3 = file_size(file_path);
+    if (file_size3 != size * 3)
+        fatal_error("File size is wrong (expected %lu got %lu)\n", size * 3, file_size3);
+    printf("file_size(%s) OK\n", file_path);
+
+    free(buf1);
+    free(buf2);
+}
+
+int main(int argc, char* argv[]) {
+    if (argc < 2)
+        fatal_error("Usage: %s <file_path>\n", argv[0]);
+
+    setup();
+    read_append(argv[1]);
+
+    printf("TEST OK\n");
+    return 0;
+}

--- a/libos/test/fs/test_fs.py
+++ b/libos/test/fs/test_fs.py
@@ -121,6 +121,16 @@ class TC_00_FileSystem(RegressionTestCase):
         self.assertIn('close(' + file_path + ') RW fd1 (mmap) OK', stdout)
         self.assertIn('close(' + file_path + ') RW fd2 OK', stdout)
 
+    def test_112_read_append(self):
+        file_path = os.path.join(self.OUTPUT_DIR, 'test_112') # new file to be created
+        stdout, stderr = self.run_binary(['read_append', file_path])
+        self.assertNotIn('ERROR: ', stderr)
+        self.assertTrue(os.path.isfile(file_path))
+        self.assertIn('TEST 1 (' + file_path + ')', stdout)
+        self.assertIn('TEST 2 (' + file_path + ')', stdout)
+        self.assertIn('TEST 3 (' + file_path + ')', stdout)
+        self.assertIn('TEST OK', stdout)
+
     # pylint: disable=too-many-arguments
     def verify_seek_tell(self, stdout, output_path_1, output_path_2, size):
         self.assertIn('Test passed', stdout)

--- a/libos/test/fs/test_tmpfs.py
+++ b/libos/test/fs/test_tmpfs.py
@@ -72,6 +72,16 @@ class TC_10_Tmpfs(test_fs.TC_00_FileSystem):
         self.assertIn('close(' + file_path + ') RW fd1 (mmap) OK', stdout)
         self.assertIn('close(' + file_path + ') RW fd2 OK', stdout)
 
+    # overrides TC_00_FileSystem to skip verification of file existence
+    def test_112_read_append(self):
+        file_path = os.path.join(self.OUTPUT_DIR, 'test_112') # new file to be created
+        stdout, stderr = self.run_binary(['read_append', file_path])
+        self.assertNotIn('ERROR: ', stderr)
+        self.assertIn('TEST 1 (' + file_path + ')', stdout)
+        self.assertIn('TEST 2 (' + file_path + ')', stdout)
+        self.assertIn('TEST 3 (' + file_path + ')', stdout)
+        self.assertIn('TEST OK', stdout)
+
     @unittest.skip("impossible to do setup on tmpfs with python only")
     def test_115_seek_tell(self):
         test_fs.TC_00_FileSystem.test_115_seek_tell(self)

--- a/libos/test/fs/tests.toml
+++ b/libos/test/fs/tests.toml
@@ -13,6 +13,7 @@ manifests = [
   "multiple_writers",
   "open_close",
   "open_flags",
+  "read_append",
   "read_write",
   "read_write_mmap",
   "seek_tell",

--- a/libos/test/ltp/ltp.cfg
+++ b/libos/test/ltp/ltp.cfg
@@ -1558,15 +1558,6 @@ skip = yes
 [pwrite01_64]
 timeout = 40
 
-[pwrite04]
-must-pass =
-    2
-
-[pwrite04_64]
-timeout = 40
-must-pass =
-    2
-
 [pwritev01]
 skip = yes
 

--- a/libos/test/ltp/ltp_sgx.cfg
+++ b/libos/test/ltp/ltp_sgx.cfg
@@ -493,11 +493,6 @@ skip = yes
 [pwrite02_64]
 skip = yes
 
-[pwrite04_64]
-timeout = 60
-must-pass =
-    2
-
 [pwritev01]
 skip = yes
 timeout = 60

